### PR TITLE
scimserver migration

### PR DESCRIPTION
**Pull request by Tirumalavasa**
related to https://github.ibm.com/IBMPrivateCloud/roadmap/issues/62411
- Fetches the mongo data from `ScimServerUsers` and `ScimServerGroups` collections
- `ScimServerUsers` data is stored into `scim_server_users` and `scim_server_users_custom` tables
-  `ScimServerGroups` data is stored into `scim_server_groups` and `scim_server_groups_custom` tables

### CODE REVIEW

File: migration/migration.go [hunk 1]
```diff
@@ -1188,4 +1190,413 @@ func removeInvalidUserPreferences(usersPrefs []v1schema.UserPreferences) (update
...
+		var ssGroupMap map[string]interface{}
+		if err = cursor.Decode(&ssGroupMap); err != nil {
+			reqLogger.Error(err, "Failed to decode Mongo document")
+			errCount++
+			err = nil
+			continue
+		}
+		ssGroupCustomMap := populateSSGroupCustomMap(ssGroupMap)
+		ssGroupCustomSchemaMap := populateSSGroupCustomSchemaMap(ssGroupCustomMap)
+		ssGroup, err = getSSGroupFromMap(ssGroupMap)
```
Comment (rwhundley) [inline migration/migration.go:1429]: Same comments as above re: handling this error.

**Review by rwhundley — CHANGES_REQUESTED** [review]

File: migration/migration.go [hunk 2]
```diff
@@ -1188,4 +1190,413 @@ func removeInvalidUserPreferences(usersPrefs []v1schema.UserPreferences) (update
...
+
+	if err = cursor.Err(); err != nil {
+		reqLogger.Error(err, "MongoDB cursor encountered an error")
+		return
+	}
+	reqLogger.Info("Successfully copied over ScimServerGroups to EDB", "docsMigrated", migrateCount)
+	return
+}
+
+func contains(slice []string, str string) bool {
```
[the remark this answers is not in the document]
Comment (Tirumalavasa) [reply migration/migration.go:1539]: Hi @rwhundley , Makes sense. let me see scan through the code on what are those 2 other places. will move it in a separate PR.

File: migration/migration.go [hunk 3]
```diff
@@ -1188,4 +1190,413 @@ func removeInvalidUserPreferences(usersPrefs []v1schema.UserPreferences) (update
...
+		if s == str {
+			return true
+		}
+	}
+	return false
+}
+
+func populateSSUserCustomMap(ssUserMap map[string]any) map[string]any {
+	if _, ok := ssUserMap["_id"]; ok {
+		delete(ssUserMap, "_id")
```
[the remark this answers is not in the document]
Comment (Tirumalavasa) [reply migration/migration.go]: Addressed it.

File: migration/migration.go [hunk 4]
```diff
@@ -1188,4 +1190,413 @@ func removeInvalidUserPreferences(usersPrefs []v1schema.UserPreferences) (update
...
+		var ssUserMap map[string]interface{}
+		if err = cursor.Decode(&ssUserMap); err != nil {
+			reqLogger.Error(err, "Failed to decode Mongo document")
+			errCount++
+			err = nil
+			continue
+		}
+		ssUserCustomMap := populateSSUserCustomMap(ssUserMap)
+		ssUserCustomSchemaMap := populateSSUserCustomSchemaMap(ssUserCustomMap)
+		ssUser, err = getSSUserFromMap(ssUserMap)
```
[the remark this answers is not in the document]
Comment (Tirumalavasa) [reply migration/migration.go:1262]: Addressed it

File: migration/migration.go (hunk 1 above)
Comment (Tirumalavasa) [reply migration/migration.go:1435]: Addressed

**Review by rashmi43 — APPROVED** [review]

**Review by rashmi43 — APPROVED** [review]

**Review by rashmi43 — COMMENTED** [review]
this might complicate it I wonder if we wanted to do anything to ensure both tables are updated for a given user.

**Pull request merged**